Introduce ClassifiedListingCategory model

citizen428 opened the pull request:
## What type of PR is this?

- [X] Refactor
- [X] Generalization

## Description

This is the first step in generalizing classified listing categories. This is **not** the full feature set, but only step 1 where we introduce a model, backfill data and make sure newly created listings will set up the correct association, while still using the old logic for the time being.

## Related Tickets & Documents

#6740

## Added tests?

- [X] yes

## Added to documentation?

- [X] no documentation needed

## Are there any post-deployment tasks we need to perform?

* [ ] Run `rake temporary:classified_listings:backfill_categories`. This should be relatively fast, as we only have ~1k listings (verified with Blazer).

citizen428 commented on app/models/classified_listing.rb:
```diff
@@ -109,4 +114,12 @@ def validate_category
+  def assign_classified_listing_category
+    category_name = CATEGORIES_AVAILABLE.dig(attributes["category"], :name)
+    category = ClassifiedListingCategory.find_by(name: category_name)
+    return unless category
```
It's "safe" to return, as the validation will fail anyway.

lightalloy replied:
I still want to add an error not to leave a user completely confused just in case :D

citizen428 replied:
There will be an error from the validation due to the `belongs_to`: `:classified_listing_category=>["must exist"]`. What else would you like to add?

lightalloy replied:
This should be enough :+1:

citizen428 commented on db/schema.rb:
```diff
@@ -308,14 +307,22 @@
     t.string "slug"
     t.string "status", default: "active"
     t.datetime "updated_at", null: false
-    t.index ["slug"], name: "index_chat_channels_on_slug", unique: true
```
I'm confused where this comes from, I'll clean it up tomorrow when I add the tests.

citizen428 commented on lib/data_update_scripts/20200409065412_add_classified_listing_categories.rb:
```diff
@@ -0,0 +1,26 @@
+module DataUpdateScripts
+  class AddClassifiedListingCategories
```
This seemed like a good idea when I wrote it and was useful for local testing, but it also alerted me to the fact that we may need to look into these scripts in the context of generalization: they are checked in and one needs to run them for the app to start up, yet some of them may be very DEV specific (like this one would add DEV categories for everyone). I'll turn this into a Rake task tomorrow that we can remove later.

lightalloy replied:
I have the same thoughts on generalization. I also think it would be useful to create categories while seeding.

rhymes replied:
Yeah, good idea, I didn't think about how DEV oriented are these categories. Should they be saved in the site config?

citizen428 replied:
I thought about SiteConfig at first, but I think having a separate model and an admin interface for it makes more sense.

lightalloy commented on db/migrate/20200409050122_add_listing_category_to_classified_listing.rb:
```diff
@@ -0,0 +1,9 @@
+  disable_ddl_transaction!
+
+  def change
+    add_reference :classified_listings,
```
I suggest adding a foreign key (`foreign_key: true`)

citizen428 replied:
Yup, good point, I usually do.

citizen428 replied:
Added, thanks for the suggestion.

rhymes commented on app/models/classified_listing_category.rb:
```diff
@@ -0,0 +1,6 @@
+  has_many :classified_listings
+
+  validates :name, :cost, :rules, presence: true
+  validates :name, uniqueness: true
```
+1 for a unique index on this one

citizen428 commented:
> I also want to namespace all the things the model but that could be controversial.

@lightalloy Do you mean `Listing::Category`? I'd generally be in favor of this, as I love namespaces, but in the past I found that that's often not super popular with other Rails developers 😢 

To make this one a bit painful I'll eventually rename the association at least, something like:

```ruby
belongs_to :category, class_name: 'ClassifiedListingCategory'
```

but I can only do that once we remove the current `category` column.

citizen428 commented:
This is now ready for review. As stated in the description, the goal is very limited for now:

1. Create classified listing categories in our DB.
2. Backfill the association for existing listings.
3. Newly created records should have the associations set up, but not use them yet.

I also marked the new association `optional: true` for now, as I don't want this to cause unnecessary problems (/cc @lightalloy in regards to our previous conversation re error messages).

The next PR will start using this model and the final one will remove the old category column etc.

rhymes commented on db/migrate/20200409043946_create_classified_listing_categories.rb:
```diff
@@ -0,0 +1,11 @@
+class CreateClassifiedListingCategories < ActiveRecord::Migration[5.2]
+  def change
+    create_table :classified_listing_categories do |t|
+      t.string :name, null: false, unique: true
```
I don't think `unique: true` is working, see the schema.rb. 

I think you need `add_index(:classified_listings_categories, :name, unique: true)` after the `create_table`

citizen428 replied:
You're absolutely right, the migration DSL is my eternal nemesis, I like plain SQL so much more. Will fix in a sec.

lightalloy commented:
I would go with smth like `ClassifiedListings::Category`.
> that's often not super popular with other Rails developers
I often use namespaces to organize the code, because otherwise the number of files grows and it's harder to work with them.

But at this point, I think it makes sense to keep the name as is in this pr. 
Overall, the pr looks good to me, besides the @rhymes's note on `unique`

citizen428 commented:
@rhymes I had waited with marking it ready for the review until the spec past, which was after I re-requested. 

@lightalloy I agree with the namespacing, like I said I used it heavily in other projects. But I'm not so keen on introducing it here now, as it's not a pattern we follow consistently, especially not for models.

The index has been fixed.

rhymes commented on db/migrate/20200409043946_create_classified_listing_categories.rb:
```diff
@@ -0,0 +1,14 @@
+class CreateClassifiedListingCategories < ActiveRecord::Migration[5.2]
+  disable_ddl_transaction!
```
no need for this as it's an empty table :)
we also should keep transactions on for `CREATE TABLE`

citizen428 replied:
The gem complained.

citizen428 replied:
> we also should keep transactions on for CREATE TABLE

I understand the general sentiment, but with the way Rails and our pipeline/deployments work I don't think we can have concurrent DDL statements anyway, right?

rhymes commented on db/migrate/20200409043946_create_classified_listing_categories.rb:
```diff
@@ -0,0 +1,14 @@
+
+      t.timestamps
+    end
+    add_index(:classified_listing_categories, :name, unique: true, algorithm: :concurrently)
```
no need for concurrently as it's an empty table :)

citizen428 replied:
The gem didn't let me do it.

citizen428 replied:
Ok, whatever the problem was was probably related to me doing/undoing/redoing these migrations a couple of times. I fixed it now.

citizen428 commented:
Sorry for requesting re-reviews, but I fixed the mgiration as suggested by rhymes and also added one more column because it makes more sense to do that here than in the follow-up PR.
